discord/http.py

```diff
@@ -260,123 +238,135 @@ async def request(
+
+                                fmt = (
+                                    "We are being rate limited. Retrying in %.2f seconds."
+                                    ' Handled under the bucket "%s"'
                                 )
-                                self._global_over.clear()
-
-                            await asyncio.sleep(retry_after)
-                            _log.debug("Done sleeping for the rate limit. Retrying...")
 
-                            # release the global lock now that the
-                            # global rate limit has passed
-                            if is_global:
-                                self._global_over.set()
-                                _log.debug("Global rate limit is now over.")
+                                # sleep a bit
+                                retry_after: float = data["retry_after"]
+                                _log.warning(fmt, retry_after, bucket_id)
+                                is_global: bool = data.get("global", False)
+
+                                if (
+                                    retry_after > self.maximum_rate_limit_time
+                                    and self.maximum_rate_limit_time != -1
+                                ):
+                                    raise HTTPException(
+                                        response,
+                                        f"Retrying rate limit would take longer than the maximum of {self.maximum_rate_limit_wait_time} seconds given",
+                                    )
+
+                                if is_global:
+                                    self.global_dynamo = DynamicBucket()
+                                    await self.global_dynamo.executed(
+                                        retry_after,
+                                        remaining or 10,
+                                        is_global=is_global,
+                                    )
+                                    self.global_dynamo = None
+                                else:
+                                    tbucket = DynamicBucket()
+                                    await self._rate_limit.push_temp_bucket(
+                                        bucket_id, tbucket
+                                    )
+                                    await tbucket.executed(
+                                        retry_after, remaining or 10, is_global=True
+                                    )
+                                    await self._rate_limit.pop_temp_bucket(bucket_id)
+
+                                _log.debug(
+                                    "Done sleeping for the rate limit. Retrying..."
+                                )
 
-                            continue
+                                continue
+
+                            # we've received a 500, 502, or 504, unconditional retry
+                            if response.status in {500, 502, 504}:
```

Comment:
I don't think we need a set here
```suggestion
                            if response.status in (500, 502, 504):
```